File: Python/Convert Sorted List to Binary Search Tree.py
# Given a singly linked list where elements are sorted in ascending order, convert it to a height balanced BST.


# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
#
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @param head, a list node
    # @return a tree node
	listNode = TreeNode(0)
	def sortedListToBST(self, head):
		leng = 0
		p = head
		while p!=None:
			leng = leng + 1
			p = p.next
		self.listNode = head
		return self.buildBST(0,leng-1)
	def buildBST(self,start,end):
		if start > end:
			return None
		mid = (start + end)/2
		leftC = self.buildBST(start,mid - 1)
		parent = TreeNode(self.listNode.val)
		parent.left = leftC
		self.listNode = self.listNode.next
		parent.right = self.buildBST(mid+1,end)
		return parent

File: Java/Convert Sorted List to Binary Search Tree.java
class Convert Sorted List to Binary Search Tree{
	private ListNode currentNode;
	public TreeNode sortedListToBST(ListNode head) {
        
    }
    private TreeNode naive(ListNode head){
    	if(head == null)
    		return null;
    	ListNode c1 = head;
    	ListNode c2 = head;
    	ListNode c3 = ListNode(0);
    	c3.next = head;
    	while(c2 != null && c2.next != null){
    		c2 = c2.next;
    		c1 = c1.next;
    		c3 = c3.next;
    	}
    	TreeNode root = new TreeNode(c1.val);
    	root.right = naive(c1.next);
    	c3.next = null;
    	if(c1 == head)
    		head = null;
    	root.left = sortedListToBST(head);
        return root;
    }
    //reconsruct from the head of list
    private TreeNode expert(ListNode head){
    	int len = 0;
        this.currentNode = head;
        ListNode p = head;
        while(p != null){
            p = p.next;
            len++;
        }
        return constructBST(0,len-1);
    }
    private TreeNode constructBST(int begin, int end){
        if (begin > end)
            return null;
        int mid = (begin + end) / 2;
        TreeNode left = this.constructBST(begin,mid-1);
        TreeNode parent = new TreeNode(this.currentNode.val);
        parent.left = left;
        this.currentNode = this.currentNode.next;
        TreeNode  right = this.constructBST(mid+1,end);
        parent.right = right;
        return parent;
    }
}